// bvsubmit/index.js
module.exports = {
  buildQuery: function(client, env, reviewData) {
  	var querystring = require('querystring');

  	function generateUserToken(userParamString) {
		var crypto = require('crypto');
	  	var sharedEncodingKey = client.encodingKey;

		var hex = new Buffer(userParamString).toString('hex');
		var md5Hash = crypto.createHash('md5').update(sharedEncodingKey + userParamString).digest('hex');

		return md5Hash + hex;
	  }
	
	var currentDate = new Date();
	currentDate = currentDate.toISOString().split('T')[0];
	var userToken = generateUserToken('userid=' + reviewData.userid + '&date='+currentDate+'&maxage=300&username='+reviewData.usernickname);
  	var postURL = "http://" + (env=="stg"?"stg.":"") + "api.bazaarvoice.com/data/submitreview.json?ApiVersion=5.4&action=submit&ProductId="+reviewData.productid+"&PassKey="+client.apiKey+"&User="+userToken;
  	return postURL;
  },
  submitQuery: function(queryString, reviewData){
  	var request = require('request');

	request.post({
	  headers: {'content-type' : 'application/x-www-form-urlencoded'},
	  url:     queryString,
	  form:    reviewData
	}, function(error, response, body){
	  console.log(error);
	  console.log(body);
	});
  }
};
